src/lib.rs

```diff
@@ -41,17 +41,30 @@ pub use plugin::PluginInitializeReturn;
 use treesitter::{get_matches, maybe_get_query};
 
 #[derive(Parser)]
+#[clap(group(
+    ArgGroup::new("query_or_filter")
+        .multiple(true)
+        .required(true)
+        .args(&["path_to_query_file", "query_source", "filter"])
+))]
 pub struct Args {
     pub paths: Vec<PathBuf>,
-    #[command(flatten)]
-    pub query_args: QueryArgs,
+    // #[arg(short = 'Q', long = "query-file", conflicts_with = "query_source",
+    // required_unless_present_any = ["query_source", "filter"])]
+    #[arg(short = 'Q', long = "query-file", conflicts_with = "query_source")]
+    pub path_to_query_file: Option<PathBuf>,
+    // #[arg(short, long, conflicts_with = "path_to_query_file", required_unless_present_any =
+    // ["path_to_query_file", "filter"])]
+    #[arg(short, long, conflicts_with = "path_to_query_file")]
+    pub query_source: Option<String>,
     #[arg(short, long = "capture")]
     pub capture_name: Option<String>,
     #[arg(short, long, value_enum)]
     pub language: Option<SupportedLanguageName>,
+    // #[arg(short, long, required_unless_present_any = ["path_to_query_file", "query_source"])]
     #[arg(short, long)]
     pub filter: Option<String>,
-    #[arg(short = 'a', long)]
+    #[arg(short = 'a', long, requires = "filter")]
```

Comment:
I should/meant to add a test for this (providing `--filter-arg` without `--filter` should be an error)